### PR TITLE
New analyzer pass to optimize in single value 

### PR DESCRIPTION
**Pull request by liuneng1994**
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
add a new analyzer pass to optimize in single value 

A new pass rule.
convert 
```sql
 SELECT * from test where x IN (1)
```
to 
```sql
SELECT * from test where x = 1
```

convert 
```sql
 SELECT * from test where x NOT IN (1)
```
to 
```sql
SELECT * from test where x != 1
```

This conversion can achieve better performance when there is only one value.

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by robot-ch-test-poll1** [comment]
*This is an automated comment for commit 532d72a7d8b74eeb1608f103e8540a7b7deb1294 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/61564/532d72a7d8b74eeb1608f103e8540a7b7deb1294/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

**Review by yariks5s — COMMENTED** [review]
There are a lot of failures regarding type mismatch and applying `equal` to an integer and a list

**Comment by liuneng1994** [comment]
image

**Comment by liuneng1994** [comment]
> There are a lot of failures regarding type mismatch and applying `equal` to an integer and a list

I added error handling logic. If resolveFunction fails, the conversion will be abandoned.

File: tests/queries/0_stateless/03013_optimize_in_to_equal.sql [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+select '-------------------';
+explain query tree select * from test where x in ('a') SETTINGS allow_experimental_analyzer = 1;
+select '-------------------';
+explain query tree select * from test where x in ('a','b') SETTINGS allow_experimental_analyzer = 1;
+select '-------------------';
+select * from test where x not in ('a') SETTINGS allow_experimental_analyzer = 1;
+select '-------------------';
+explain query tree select * from test where x not in ('a') SETTINGS allow_experimental_analyzer = 1;
+select '-------------------';
+select * from test where x not in (NULL) SETTINGS allow_experimental_analyzer = 1;
```
Comment (yariks5s) [inline tests/queries/0_stateless/03013_optimize_in_to_equal.sql:16]: also good to check `explain query` for this query and other cases, that do not support this conversion (x in array and x in tuple)

**Review by yariks5s — COMMENTED** [review]
good in general, described my thoughts below

File: src/Functions/equals.h [hunk 2]
```diff
@@ -0,0 +1,11 @@
+#pragma once
```
[the remark this answers is not in the document]
Comment (liuneng1994) [reply src/Functions/equals.h:1]: I will also use it for other subsequent optimizations, which can be better reused through .h

File: tests/queries/0_stateless/03013_optimize_in_to_equal.sql (hunk 1 above)
Comment (liuneng1994) [reply tests/queries/0_stateless/03013_optimize_in_to_equal.sql]: done

**Comment by yariks5s** [comment]
`03010_read_system_parts_table_test` is currently broken, will reload CI once it's fixed

**Review by yariks5s — APPROVED** [review]

**Comment by Algunenano** [comment]
This improvement has introduced many problems:
* Some where fixed in https://github.com/ClickHouse/ClickHouse/pull/62066
* Some still present https://github.com/ClickHouse/ClickHouse/issues/62129
* Most importantly, there isn't a setting controlling this change, so it's impossible to disable it.

I'm reverting it and removing it from 23.3. Please reintroduce it with proper fixes and settings.

**Comment by liuneng1994** [comment]
> * , there isn't a setting controlling this ch

I will fix it, and add a new settings

**Pull request merged**